xonsh/events.py

```diff
@@ -0,0 +1,113 @@
+"""
+Events for xonsh.
+
+In all likelihood, you want builtins.__xonsh_events__
+
+The best way to "declare" an event is something like::
+
+    __xonsh_events__.on_spam.__doc__ = "Comes with eggs"
+"""
+
+
+class Event(set):
+    """
+    A given event that handlers can register against.
+
+    Acts as a ``set`` for registered handlers.
+
+    Note that ordering is never guarenteed.
+    """
+    def __init__(self, doc=None):
+        self.__doc__ = doc
+
+    def handler(self, callable):
+        """
+        Registers a handler. It's suggested to use this as a decorator.
+        """
+        self.add(callable)
+        return callable
+
+    def calleach(self, *pargs, **kwargs):
+        """
+        The core handler caller that all others build on.
+
+        This works as a generator. Each handler is called in turn and its
+        results are yielded.
+
+        If the generator is interupted, no further handlers are called.
+
+        The caller may send() new positional arguments (eg, to implement
+        modifying semantics). Keyword arguments cannot be modified this way.
+        """
+        for handler in self:
+            newargs = yield handler(*pargs, **kwargs)
```

Comment:
The way I figured is that each subsystem would have its own event registrations, so what order handlers are registered depends on module load order, which is either import order or what order amalgamate decides. I would call this deterministic but chaotic, and not easily reasoned about.

Which makes me think that for the sake of future sanity, any kind of event semantics that depend on ordering should be avoided. That is, events could be no responses, one response, or many responses. The primary use case for chained responses--preprocessors--is problematic anyway because each preprocessor has to be able to deal with arbitrary unknown syntax because it doesn't know what other preprocessors might be installed on the system.